### ISSUE TITLE
handshake failed (The TLS connection was non-properly terminated.)

### ISSUE THREAD

**Issue by InputOutputZ** [PowerDNS/pdns#10978]
- Program: dnsdist 
 - Issue type: Failing to establish TLS 1.3 connection for both DoT and DoH.

### Short description
Using kdig, I fail to establish successful connection over DoH and DoT when minTLSVersion set to 'tls1.3', for DoH returns " TLS, handshake failed (The TLS connection was non-properly terminated.)" and for DoT returns " TLS, handshake failed (A TLS fatal alert has been received.) "

### Environment
 - Operating system: centOS 7
 - Software version: dnsdist-1.7.0-alpha2, openssl 3.1.0-dev
 - Software source: PDNS official git repository.

### Steps to reproduce

Configuration options for addTLSLocal and addDOHLocal

`{ minTLSVersion='tls1.3', reusePort=true, tcpFastOpenQueueSize=3, ciphers='AES128-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384', ciphersTLS13='TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:TLS_AES_128_CCM_SHA256:TLS_AES_128_CCM_8_SHA256'})`

Steps for DoT
1- `kdig @127.0.0.10 [website.offering.dot.over.853] +tls`

Steps for DoH
1- `kdig -p 443 @127.0.0.10 [doh.website.com] +https`

### Expected behaviour
Instead of TLS 1.2 it should be establishing TLS session with TLS 1.3 and returning DNS query answer like following with status noerror
```
;; TLS session (TLS1.2)-(ECDHE-ECDSA-SECP256R1)-(AES-128-GCM)
;; ->>HEADER<<- opcode: QUERY; status: NOERROR; id: 54651
;; Flags: qr rd ra; QUERY: 1; ANSWER: 1; AUTHORITY: 0; ADDITIONAL: 1

;; EDNS PSEUDOSECTION:
;; Version: 0; flags: ; UDP size: 512 B; ext-rcode: NOERROR

;; QUESTION SECTION:
;; website.com.    		IN	A

;; ANSWER SECTION:
website.com.    	13902	IN	A	213.171.210.111

;; Received 60 B
;; Time 2021-11-13 15:41:23 UTC
;; From 127.0.0.10@853(TCP) in 1.9 ms
```

### Actual behaviour
DoT
```
;; WARNING: TLS, handshake failed (A TLS fatal alert has been received.)
;; ERROR: failed to query server 127.0.0.10@853(TCP)`

DoH
`;; WARNING: TLS, handshake failed (The TLS connection was non-properly terminated.)
;; ERROR: failed to query server 127.0.0.10@443(TCP)
```

### Other information
I'm not sure if its actually a bug or misconfiguration from my side, given I was able to start TLS handshake with openssl using tls 1.3 at both DoH and DoT ports yet couldn't figure out how to test the connection termination, and failed to conclude if its actually a bug with kdig only, yet I've found this bug https://github.com/PowerDNS/pdns/issues/7716 and it seemed handling ciphersuites for TLS 1.3 from dnsdist side isn't immune to bugs. Please advise me if there is actually a bug with one of the ciphersuites I used or kdig is buggy, although I've seen other people were able to instantiate successful TLS 1.3 connection for both DoH and DoT using kdig.

**Comment by rgacogne** [comment]
Sadly CentOS 7 ships OpenSSL `1.0.2k` which does not support TLS 1.3, so you would need to install a more recent version of OpenSSL on your system and recompile dnsdist against that version.
You should be able to confirm that by looking at the counters in `showTLSErrorCounters()` after trying to open a TLS 1.3 connection to dnsdist.

**Comment by InputOutputZ** [comment]
> Sadly CentOS 7 ships OpenSSL `1.0.2k` which does not support TLS 1.3, so you would need to install a more recent version of OpenSSL on your system and recompile dnsdist against that version.
> You should be able to confirm that by looking at the counters in `showTLSErrorCounters()` after trying to open a TLS 1.3 connection to dnsdist.

Oh, sorry I forgot to mention, that I compiled dnsdist with openssl 3.1.0-dev. I've TLS 1.3 enabled and supported?

**Comment by rgacogne** [comment]
Ah, right, then in theory that should work. I'm afraid we did not do any testing with OpenSSL 3.x, though. Did you get the change to look at the counters I mentioned earlier? Getting a network capture of the actual TLS exchanges would likely be useful as well.

**Comment by InputOutputZ** [comment]
> Ah, right, then in theory that should work. I'm afraid we did not do any testing with OpenSSL 3.x, though. Did you get the change to look at the counters I mentioned earlier? Getting a network capture of the actual TLS exchanges would likely be useful as well.

I've just tried to capture the exchanges but couldn't figure out where dnsdist outputs the capture?

**Comment by rgacogne** [comment]
Uh, I'm not sure I understand what you mean, dnsdist doesn't do network capture by itself. If you mean `showTLSErrorCounters()`, you need to set up a console configuration, connect to the running dnsdist using a console connection and execute that command: https://dnsdist.org/guides/console.html

**Comment by InputOutputZ** [comment]
> Uh, I'm not sure I understand what you mean, dnsdist doesn't do network capture by itself. If you mean `showTLSErrorCounters()`, you need to set up a console configuration, connect to the running dnsdist using a console connection and execute that command: https://dnsdist.org/guides/console.html

This is the results coming up, unsupported protocol for both DoT and DoH, using kdig.

`> showTLSErrorCounters()
#   Address              DH key too small        Inappropriate fallback  No shared cipher        Unknown cipher type     Unknown exchange type   Unknown protocol        Unsupported EC          Unsupported protocol   
0   0.0.0.0:6565         0                       0                       0                       0                       0                       0                       0                       2                      
1   [::]:6565            0                       0                       0                       0                       0                       0                       0                       0                      
2   0.0.0.0:853          0                       0                       0                       0                       0                       0                       0                       1                      
3   [::]:853             0                       0                       0                       0                       0                       0                       0                       0 `

when I attempt to establish connection using openssl e.g with

`openssl s_client -tls1_3 -connect website.com:853`

I get 

`Connecting to 127.0.0.1
CONNECTED(00000003)
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: ECDSA
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 4362 bytes and written 329 bytes
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 256 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
....
`

It shows TLS 1.3 is working? and at DoH port, I get TLS 1.3 working fine on browser?

**Comment by rgacogne** [comment]
> It shows TLS 1.3 is working? and at DoH port, I get TLS 1.3 working fine on browser?

Yes, it looks like it. The fact that `Unsupported protocol` increases when you try to connect with `kdig` indicates that we got a `SSL_R_UNSUPPORTED_PROTOCOL` or `SSL_R_VERSION_TOO_LOW` from OpenSSL. Given that you can connect just fine with `openssl`, it might indicate that `kdig` is trying to connect using TLS 1.2. Are you sure that your version of `kdig` supports and is using TLS 1.3?

**Comment by InputOutputZ** [comment]
`yum install yum-plugin-copr
yum install knot-utils`

This is how I installed kdig (Knot DNS), version 3.1.3 in centos 7, and I think it supports TLS 1.3, as it was released just a month ago, per this https://github.com/CZ-NIC/knot/tags.

It seems that its more to do with kdig, maybe conflicting with the old openssl in centos 7 which I noticed usually along with librassl older version is used and referenced by e.g. yum package, and had to keep it in system in order to keep yum working.

Or maybe, the absence of valid OCSP stapling response could cause this failure at kdig end? given I've only tested so far with haproxy as frontend and yet haven't set it with an OCSP response?, refer to https://github.com/CZ-NIC/knot/commit/384d393ea834c2c847b6af3c3c30b970bd877a96#diff-74e44e1557ec6120b7be114be7f6f06987ae916a7d2ea60a3f19f1689a2c9aa1

Anyhow, if you deem this isn't a bug, there is no problem from my side for this case to be closed, with thanks for your elaboration on how to setup the dnsdist console.

**Comment by InputOutputZ** [comment]
> > It shows TLS 1.3 is working? and at DoH port, I get TLS 1.3 working fine on browser?
> 
> Yes, it looks like it. The fact that `Unsupported protocol` increases when you try to connect with `kdig` indicates that we got a `SSL_R_UNSUPPORTED_PROTOCOL` or `SSL_R_VERSION_TOO_LOW` from OpenSSL. Given that you can connect just fine with `openssl`, it might indicate that `kdig` is trying to connect using TLS 1.2. Are you sure that your version of `kdig` supports and is using TLS 1.3?

```
yum install yum-plugin-copr
yum install knot-utils
```

This is how I installed kdig (Knot DNS), version 3.1.3 in centos 7, and I think it supports TLS 1.3, as it was released just a month ago, per this https://github.com/CZ-NIC/knot/tags.

It seems that its more to do with kdig, maybe conflicting with the old openssl in centos 7 which I noticed usually along with librassl older version is used and referenced by e.g. yum package, and had to keep it in system in order to keep yum working.

Or maybe, the absence of valid OCSP stapling response could cause this failure at kdig end? given I've only tested so far with haproxy as frontend and yet haven't set it with an OCSP response?, refer to https://github.com/CZ-NIC/knot/commit/384d393ea834c2c847b6af3c3c30b970bd877a96#diff-74e44e1557ec6120b7be114be7f6f06987ae916a7d2ea60a3f19f1689a2c9aa1

Anyhow, if you deem this isn't a bug, there is no problem from my side for this case to be closed, with thanks for your elaboration on how to setup the dnsdist console.

**Comment by rgacogne** [comment]
It's very unlikely that OCSP plays a role, here, as I can connect just fine with `kdig` to a dnsdist with no OCSP stapling.
Can you try if your `kdig` from CentOS 7 can connect to a public dnsdist instance, like quad9, using TLS 1.3? This works for me using Arch Linux's `kdig` 3.1.4:
```
$ kdig @9.9.9.9:853 +tls powerdns.com
;; TLS session (TLS1.3)-(ECDHE-SECP256R1)-(ECDSA-SECP256R1-SHA256)-(AES-256-GCM)
```

**Comment by InputOutputZ** [comment]
> dig @9.9.9.9:853 +tls powerdns.com

```
;; TLS session (TLS1.2)-(ECDHE-ECDSA-SECP256R1)-(AES-256-GCM)
;; ->>HEADER<<- opcode: QUERY; status: NOERROR; id: 21156
;; Flags: qr rd ra; QUERY: 1; ANSWER: 1; AUTHORITY: 0; ADDITIONAL: 1

;; EDNS PSEUDOSECTION:
;; Version: 0; flags: ; UDP size: 1232 B; ext-rcode: NOERROR

;; QUESTION SECTION:
;; powerdns.com.       		IN	A

;; ANSWER SECTION:
powerdns.com.       	3600	IN	A	188.166.104.92

;; Received 57 B
;; Time 2021-11-16 17:32:47 UTC
;; From 9.9.9.9@853(TCP) in 15.7 ms

```

With quad9 TLS 1.3 is not set as the minimum version, it connects with TLS 1.2,y default, that is the same outcome with my installation, the issue happens when I set min version to 1.3 in dnsdist config? have you tried this with your dnsdist installation? which ciphersuites you've used then?

**Comment by InputOutputZ** [comment]
> > kdig @9.9.9.9:853 +tls powerdns.com
> 
> ```
> ;; TLS session (TLS1.2)-(ECDHE-ECDSA-SECP256R1)-(AES-256-GCM)
> ;; ->>HEADER<<- opcode: QUERY; status: NOERROR; id: 21156
> ;; Flags: qr rd ra; QUERY: 1; ANSWER: 1; AUTHORITY: 0; ADDITIONAL: 1
> 
> ;; EDNS PSEUDOSECTION:
> ;; Version: 0; flags: ; UDP size: 1232 B; ext-rcode: NOERROR
> 
> ;; QUESTION SECTION:
> ;; powerdns.com.       		IN	A
> 
> ;; ANSWER SECTION:
> powerdns.com.       	3600	IN	A	188.166.104.92
> 
> ;; Received 57 B
> ;; Time 2021-11-16 17:32:47 UTC
> ;; From 9.9.9.9@853(TCP) in 15.7 ms
> ```
> 
> With quad9 TLS 1.3 is not set as the minimum version, it connects with TLS 1.2,y default, that is the same outcome with my installation, the issue happens when I set min version to 1.3 in dnsdist config? have you tried this with your dnsdist installation? which ciphersuites you've used then?

**Comment by rgacogne** [comment]
I stated above I'm using  `kdig` 3.1.4 from Arch Linux, not CentOS 7, so with GnuTLS 3.7.2. If you are using the GnuTLS library shipped in CentOS 7, which is 3.3.x, you are not going to get TLS 1.3 support in `kdig`, as that requires GnuTLS >= 3.6.
If you really want TLS 1.3 support I would recommend moving away from CentOS 7 and upgrading to a newer distribution, as otherwise you will have to recompile pretty much all TLS libraries and the binaries that use them.

**Comment by rgacogne** [comment]
I'm closing this as I believe this is not an issue in dnsdist.
